feat: 프로젝트 기본 셋팅 (#1)

kaley0421 opened the pull request:
## 📍 이슈 번호 
- close #1

## 📝 작업 내용
* spring boot 프로젝트를 셋팅합니다.
     - 프로젝트 생성
     - 공통 응답 처리
     - 예외처리 기능
     - Base Entity 생성
     - swagger 세팅
     - QueryDsl 세팅

## 📸 스크린샷 

## 🔍 pr 포인트 혹은 궁금한 점

kaley0421 committed:
feat: BaseResponse 생성 (#1)

kaley0421 committed:
feat: BaseTimeEntity 생성 (#1)

kaley0421 committed:
feat: 예외처리 기능 구현 (#1)

kaley0421 committed:
feat: swagger 추가 (#1)

kaley0421 committed:
feat: security 설정 추가 (#1)

kaley0421 committed:
feat: queryDsl 추가 (#1)

Mingyum-Kim commented on build.gradle:
```diff
@@ -31,8 +31,21 @@ dependencies {
 }
+
+clean {
+	delete file ('src/main/generated')
```
이거는 Querydsl에서 Q파일이 생성되어서 Conflict 생기는 것 때문에 추가한 것 맞나요?

kaley0421 replied:
넵 맞습니다!!

Mingyum-Kim commented on src/main/java/com/dnd/favolink/global/common/entity/BaseTimeEntity.java:
```diff
@@ -0,0 +1,24 @@
+
+@Getter
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
```
이거는 JPA Auditing 기능을 사용하기 위한 설정 어노테이션이군요 😃
제가 알기로는 Auditing을 사용하려면 어플리케이션 클래스에 어노테이션을 하나 더 붙여야하는데 확인 부탁드려요!

kaley0421 replied:
`@EnableJpaAuditing` 말씀하시는거 맞으실까요?!?
추가해주는걸 까먹었네요..😅🥲 수정하도록 하겠습니다!!

Mingyum-Kim commented on src/main/java/com/dnd/favolink/global/common/response/BaseResponse.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.http.ResponseEntity;
+
+public class BaseResponse {
+    public static <T> ResponseEntity<T> ok(final T data) {
+        return ResponseEntity.ok().body(data);
+    }
```
컨트롤러 작성할 때는 이런 형식으로 작성하면 될까요? 
```java
@GetMapping("/post")
public ResponseEntity<Post> getPost(int id)  {
   return new ResponseEntity<>(PostService.getPost(.....));
}
```

kaley0421 replied:
```java
@GetMapping("/post")
public ResponseEntity<Post> getPost(int id)  {
   return new BaseResponse.ok(PostService.getPost(.....));
}
```
요런식으로 작성한다고 생각했습니다!!
BaseResponse 라는 클래스를 굳이 사용해준 이유는 추후에 '응답 데이터 형식을 추가하고 싶은 경우' 에 다음과 같이 함수를 추가하여 정형화 해주기에 용이할 것 같다고 생각해서였습니다!
ex.
```java
 public static <T> ResponseEntity<T> created(T data) {    // 데이터 생성 성공 시 응답형식을 '201 created + 생성된 데이터' 로 통일하고 싶은 경우
        return ResponseEntity
                .status(HttpStatus.CREATED)
                .body(data);
}

public static ResponseEntity<Void> noContent() {   // 요청은 성공했으나, 반환할 데이터가 딱히 없는 경우
    return ResponseEntity
            .status(HttpStatus.NO_CONTENT)
            .build();
}
  ```

Mingyum-Kim replied:
네 확인했습니다! 정정 감사합니다.

Mingyum-Kim commented on src/main/java/com/dnd/favolink/global/common/response/ErrorResponse.java:
```diff
@@ -0,0 +1,108 @@
+    @Getter
+    @NoArgsConstructor(access = AccessLevel.PROTECTED)
+    public static class CustomFieldError {
+        private String field;
+        private String value;
+        private String reason;
```
갑자기 궁금해졌는데 이 객체를 전송할 때는 'reason'과 같이 일관성이 필요한 필드는 어떻게 통일할 수 있을까요? 
예를 들어서 같은 빈 문자열 입력이어도 "빈 문자열이 입력되었습니다" "문자를 입력해주세요" 등 사람마다 다양하게 이유를 적을 수 있는데, 케이스 별로 에러 메시지 (혹은 이유)를 통일하면 좋을 것 같아서요!

kaley0421 replied:
음.. reason 에 들어가는 값들이 모두 `Exception 객체`(ex. `BindException`, `MethodArgumentTypeMismatchException`, `NullPointerException`...) 같은 기본 exception 객체들에서 불러오는 에러메시지 값들이라, 이를 어떻게 정형화 시켜야 할지는 잘 떠오르지 않는것 같아요..! 🥲
ex.
```json
{
    ...
    "message": "입력 파라미터가 필요합니다.",       // 직접 작성하는 에러메시지 (ErrorCode Enum 으로 지정)
    "errors": [
        {
            "field": "id",
            "value": "",
            // 발생한 exception 객체에서 불러온 에러메시지
            "reason": "Required request parameter 'id' for method parameter type Long is not present"  
        }
    ]
}
```


다만 말씀해주신 예시처럼 같은 '저희가 직접 작성하게되는 에러메시지' 들에 대해서는 `CommonErrorCode` enum 에 `에러코드 & 에러메시지` 를 정의해 두면 통일할 수 있을 것 같습니다!!!

Mingyum-Kim replied:
아 저는 reason이 개발자가 직접 작성하는 메시지가 들어가는 것으로 오해하였습니다!
Enum으로 에러 메시지를 정의하는 것도 좋다고 생각합니다. 감사합니다 :)

Mingyum-Kim commented on src/main/java/com/dnd/favolink/global/config/SecurityConfig.java:
아 SecurityConfig도 같이 추가해주셨군요 !! 감사합니다 👍😊

Mingyum-Kim commented on src/main/java/com/dnd/favolink/global/config/SwaggerConfig.java:
깔꼼합니다!

Mingyum-Kim commented on src/main/java/com/dnd/favolink/global/error/code/CommonErrorCode.java:
```diff
@@ -0,0 +1,24 @@
+
+@Getter
+@AllArgsConstructor
+public enum CommonErrorCode implements ErrorCode {
+    INTERNAL_SERVER_ERROR(HttpStatus.INTERNAL_SERVER_ERROR,"C-000","서버 내부 오류가 발생하였습니다."),
+    BAD_REQUEST(HttpStatus.BAD_REQUEST,"C-001", "잘못된 요청입니다."),
+    INVALID_INPUT_VALUE(HttpStatus.BAD_REQUEST, "C-002", "유효하지 않은 입력입니다."),
+    TYPE_MISMATCH(HttpStatus.BAD_REQUEST, "C-003", "입력 타입이 일치하지 않습니다."),
+    MISSING_REQUEST_PARAMS(HttpStatus.BAD_REQUEST, "C-004", "입력 파라미터가 필요합니다."),
+    UNAUTHORIZED(HttpStatus.UNAUTHORIZED,"C-005", "로그인이 필요합니다."),
+    NOT_FOUND(HttpStatus.NOT_FOUND,"C-006","리소스를 찾을 수 없습니다."),
+    METHOD_NOT_ALLOWED(HttpStatus.METHOD_NOT_ALLOWED, "C-007", "허용되지 않은 http 메소드입니다."),
+    FORBIDDEN(HttpStatus.FORBIDDEN,"C-008","접근이 허용되지 않았습니다."),
+    NULL_POINT(HttpStatus.NOT_FOUND, "C-009", "null pointer exception 발생");
```
C-003 과 같은 코드 네이밍은 어떻게 하면 되는건가요?
C는 코드의 약자이겠죠 ,,, ?? 😶

kaley0421 replied:
사실 코드 네이밍은 제가 혼자서 정한 건데, 저는 Common 의 약자라고 생각했습니다..! 😂
예를 들어 User 도메인과 관련하여 발생하는 에러라면
```java
public enum UserErrorCode implements ErrorCode {
    USER_NOT_EXIST(HttpStatus.BAD_REQUEST, "U-001", "존재하지 않는 유저입니다.");
    ...
}
```
이런 식으로 도메인의 초성(?) 알파벳을 따서 만든다고 생각했습니다!
에러코드 네이밍 관련해서는 더 좋은 방법이 있을지 한번 논의해봐도 좋을 것 같아요!!

Mingyum-Kim replied:
아 그런 방식이 일반적으로 많이 사용되는 건가요? (잘 몰라서 여쭤봅니다!)
알파벳이 중복될 수도 있고 ,, 그런 생각이 조금 드네요! 다시 얘기해보면 좋을 것 같아요 :)

Mingyum-Kim commented on src/main/java/com/dnd/favolink/global/error/exception/BusinessException.java:
```diff
@@ -0,0 +1,30 @@
+    private ErrorCode errorCode;
+    private List<ErrorResponse.CustomFieldError> errors = new ArrayList<>();
+
+    public BusinessException(ErrorCode errorCode, String message) {
+        super(message);
+        this.errorCode = errorCode;
+    }
```
```java
throw new BusinessException(ErrorCode.MEMBER_NOT_FOUND, "사용자를 찾을 수 없습니다."); 
```
이런 식으로 작성하면 되겠네요!

kaley0421 replied:
네네 맞습니다!
저는 ErrorCode enum 을 만들면서 'api 결과로 내려주고 싶은 오류메시지' 를 적고 BusinessException 에는 ErrorCode 만 담아서 전달하는 방식을 생각했습니다! 
ex. 아래의 경우 USER_NOT_EXIST 에 정의되어있는 <http 상태 + U-001 이라는 코드 + "존재하지 않는 유저입니다." 메시지> 가 API 응답으로 떨어집니다!
```java
public enum UserErrorCode implements ErrorCode {     // 에러코드 정의
    USER_NOT_EXIST(HttpStatus.BAD_REQUEST, "U-001", "존재하지 않는 유저입니다.");
    ...
}
```
```java
throw new BusinessException(UserErrorCode.USER_NOT_EXIST);   // throw error
```


민겸님이 작성해주신 것처럼 BusinessException 생성 시 'ErrorCode 와 메시지'를 함께 전달할 경우,
- 에러 로그에는 BusinessException 생성 시 작성해준 메시지가 찍히고,
- API 응답 결과로는 ErrorCode 에 작성해둔 메시지로 응답하게 구현되어 있습니다!
```java
throw new BusinessException(CommonErrorCode.BAD_REQUEST, CommonErrorCode.BAD_REQUEST.getMessage() + " | userid : " + "jimin");
```
요런 식으로 보다 세부적인 로그를 남기고 싶을 때 사용하면 좋을 것 같습니다..!

Mingyum-Kim replied:
아하 그러면 BusinessException에는 조금 더 개발자 친화적으로, ErrorCode에는 사용자 친화적으로 메시지를 작성하면 되겠군요! 세세한 점까지 고려해서 작성해주신 것 같아요. 설명 감사합니다 :)

Mingyum-Kim reviewed (commented):
지민님 ! 꼼꼼하게 공통 응답 클래스부터 예외 핸들러까지 잘 만들어주셨네요 😊
고생하셨습니다 궁금한 점 등을 적어서 리뷰 남겨놓았으니 확인 부탁드려요 !

kaley0421 committed:
feat: jpa auditing 어노테이션 추가 (#1)

kaley0421 committed:
fix: ErrorResponse, GlobalExceptionHandler 수정 (#1)
- ConstraintViolationException, MethodArgumentTypeMismatchException, MissingServletRequestParameterException 처리 관련

kaley0421 committed:
feat: security config 에 health-check api 추가 (#1)

kaley0421 committed:
Merge branch 'dev' into feat/1
